nf_core/pipelines/download.py

```diff
@@ -1281,16 +1294,33 @@ def singularity_image_filenames(self, container: str) -> Tuple[str, Optional[str
         elif self.container_cache_utilisation in ["amend", "copy"]:
             raise FileNotFoundError("Singularity cache is required but no '$NXF_SINGULARITY_CACHEDIR' set!")
 
-        return (out_path, cache_path)
+        library_path = None
+        if os.environ.get("NXF_SINGULARITY_LIBRARYDIR"):
+            library_path = os.path.join(os.environ["NXF_SINGULARITY_LIBRARYDIR"], out_name)
 
-    def singularity_copy_cache_image(self, container: str, out_path: str, cache_path: Optional[str]) -> None:
+        return (out_path, cache_path, library_path)
+
+    def singularity_copy_cache_image(self, container: str, out_path: str, cache_path: str) -> None:
         """Copy Singularity image from NXF_SINGULARITY_CACHEDIR to target folder."""
-        # Copy to destination folder if we have a cached version
-        if cache_path and os.path.exists(cache_path):
-            log.debug(f"Copying {container} from cache: '{os.path.basename(out_path)}'")
-            shutil.copyfile(cache_path, out_path)
-            # Create symlinks to ensure that the images are found even with different registries being used.
-            self.symlink_singularity_images(out_path)
+        self.singularity_copy_image(container, cache_path, out_path)
+        # Create symlinks to ensure that the images are found even with different registries being used.
+        self.symlink_singularity_images(cache_path)
+
+    def singularity_copy_library_image(
+        self, container: str, library_path: str, out_path: str, cache_path: Optional[str]
+    ) -> None:
+        """Copy Singularity image from NXF_SINGULARITY_LIBRARYDIR to target folder, and possibly NXF_SINGULARITY_CACHEDIR."""
+        self.singularity_copy_image(container, library_path, out_path)
+        if cache_path:
+            self.singularity_copy_image(container, library_path, cache_path)
+
+    def singularity_copy_image(self, container: str, from_path: str, to_path: str) -> None:
+        """Copy Singularity image between folders. This function is used seamlessly
+        across the target directory, NXF_SINGULARITY_CACHEDIR, and NXF_SINGULARITY_LIBRARYDIR."""
+        log.debug(f"Copying {container} to cache: '{os.path.basename(from_path)}'")
+        shutil.copyfile(from_path, to_path)
```

Comment:
Sorry, but this function misses important functionality. If you factor out the copying process, you need to consider that copies may be interrupted by exceptions or by the user using SIGINT (CTRL+C). 

Previously, in case of incomplete/corrupted downloads, the local files were deleted by the `except` and `finally` branches within the `singularity_download_image()` function. (Lines 1[356-1369](https://github.com/nf-core/tools/blob/9e0557447fdd967f54798da0116ef2f9f2015535/nf_core/pipelines/download.py#L1356-1369) on dev). Additionally, it did not matter too much, since the copy happened from the cache to the `output_path`, typically a folder the user would delete in case of download failures. 

But now that you changed the logic, it seems more likely to me that a user could amass corrupted images in the cache folder persistently. Therefore, it is important to ensure that the copy process removes partially copied files etc. in case of exceptions or SIGINT, e.g. with a `cleanup_temp_files()` function. 

Something along this line:

```Python
import signal
import sys

# Example
def cleanup_temp_files():
    if os.path.exists('temp_file'):
        os.remove('temp_file')

# Define a signal handler for SIGINT (CTRL+C)
def abort_download(sig, frame):
    cleanup_temp_files()
    raise DownloadError("Aborting pipeline download due to user interruption.")

signal.signal(signal.SIGINT, abort_download)

# Example of using try-finally for cleanup in case of exceptions
try:
    # File copy code here
    # ...
except Exception as e:
    cleanup_temp_files()
      raise DownloadError(e) from e
finally:
    cleanup_temp_files()
```